modules/custom/social_graphql/modules/social_graphql_oauth2/src/Entity/Server.php

```diff
@@ -0,0 +1,167 @@
+   *   The resolve info object.
+   *
+   * @return array
+   *   Allow directive values array.
+   */
+  private function getAllowDirectiveValues(ResolveInfo $info): array {
+    $allow_directives = [
+      'allowUser',
+      'allowBot',
+      'allowAll',
+    ];
+    $directive_values = [];
+
+    foreach ($allow_directives as $allow_directive) {
+      $directive_def = $info->schema->getDirective($allow_directive);
+
+      if ($directive_def !== NULL) {
+        $type = $info->returnType;
+        if ($type instanceof WrappingType) {
+          $type = $type->getWrappedType(TRUE);
+        }
+
+        // Get directive values on the type.
+        $type_directive_values = $type->astNode !== NULL ? Values::getDirectiveValues(
+          $directive_def,
+          $type->astNode
+        ) : NULL;
+
+        // Get directive values on the field.
+        $field_directive_values = $info->fieldDefinition->astNode ? Values::getDirectiveValues(
+          $directive_def,
+          $info->fieldDefinition->astNode
+        ) : NULL;
+
+        $values = $type_directive_values ?: $field_directive_values;
+
+        if (!empty($values)) {
+          $directive_values[$allow_directive] = $values;
+        }
+      }
+    }
+
+    return $directive_values;
+  }
+
+  /**
+   * Checks access based on the directive values and token.
+   *
+   * @param array $directive_values
+   *   The directive values.
+   * @param \Drupal\simple_oauth\Entity\Oauth2TokenInterface $token
+   *   The OAuth2 token.
+   */
+  private function checkAccess(array $directive_values, Oauth2TokenInterface $token): void {
+    $token_has_user = !$token->get('auth_user_id')->isEmpty();
+    /** @var \Drupal\simple_oauth\Plugin\Field\FieldType\Oauth2ScopeReferenceItemListInterface $scopes_field */
+    $scopes_field = $token->get('scopes');
+    $token_scopes = array_map(function (Oauth2ScopeInterface $scope) {
+      return $scope->getName();
+    }, $scopes_field->getScopes());
+
+    $directive_name = $token_has_user ? 'allowUser' : 'allowBot';
+    $grant_type = $token_has_user ? 'client credentials' : 'authorization code';
+    $required_scopes = $this->getRequiredScopes($directive_name, $directive_values);
+    if (empty($required_scopes)) {
+      throw new UserError(sprintf("The '%s' grant type is required.", $grant_type));
+    }
+    $this->checkRequiredScopes($required_scopes, $token_scopes);
+  }
+
+  /**
+   * Checks if the access token has the scope id's.
+   *
+   * @param array $required_scopes
+   *   The required scope id's.
+   * @param array $token_scopes
+   *   The scopes on the token.
+   *
+   * @throws \GraphQL\Error\UserError
+   */
+  private function checkRequiredScopes(array $required_scopes, array $token_scopes): void {
+    foreach ($required_scopes as $scope) {
+      if (!in_array($scope, $token_scopes)) {
+        throw new UserError(sprintf("The '%s' scope is required.", $scope));
+      }
+    }
```

Comment:
Do we think it'd be a DX improvement to not exit early but report all missing scopes? Or do we think that's difficult because there may be different required scopes for `allowAll` or `allowType` where only requesting a single scope could solve all errors?

We could also log this as future improvement.

Reply:
The UserError only accepts a single string, it would be perhaps messy to communicate all errors in a single string. I agree we could indeed log this as future improvement.